code/modules/clothing/chameleon/chameleon_action_subtypes.dm

```diff
@@ -0,0 +1,293 @@
+		agent_card.assigned_icon_state = initial(copied_advanced_card.assigned_icon_state)
+
+	agent_card.update_label()
+	agent_card.update_appearance(UPDATE_ICON)
+
+/datum/action/item_action/chameleon/change/id/apply_job_data(datum/job/job_datum)
+	var/obj/item/card/id/advanced/chameleon/agent_card = target
+	agent_card.forged = TRUE
+
+	// job_outfit is going to be a path.
+	var/datum/outfit/job/job_outfit = job_datum.outfit
+	if(!job_outfit)
+		return
+
+	// copied_card is also going to be a path.
+	var/obj/item/card/id/copied_card = initial(job_outfit.id)
+	if(!copied_card)
+		return
+
+	// If the outfit comes with a special trim override, we'll use that. Otherwise, use the card's default trim. Failing that, no trim at all.
+	var/new_trim = initial(job_outfit.id_trim) ? initial(job_outfit.id_trim) : initial(copied_card.trim)
+
+	if(new_trim)
+		SSid_access.apply_trim_to_chameleon_card(agent_card, new_trim, FALSE)
+	else
+		agent_card.assignment = job_datum.title
+
+	agent_card.icon_state = initial(copied_card.icon_state)
+	if(ispath(copied_card, /obj/item/card/id/advanced))
+		var/obj/item/card/id/advanced/copied_advanced_card = copied_card
+		agent_card.assigned_icon_state = initial(copied_advanced_card.assigned_icon_state)
+
+	agent_card.update_label()
+	agent_card.update_appearance(UPDATE_ICON)
+
+/datum/action/item_action/chameleon/change/id_trim
+	chameleon_type = /datum/id_trim
+	chameleon_name = "ID Trim"
+
+/datum/action/item_action/chameleon/change/id_trim/New(Target)
+	. = ..()
+	if(!istype(target, /obj/item/card/id/advanced/chameleon))
+		stack_trace("Adding chameleon ID trim action to non-chameleon id ([target])")
+		qdel(src)
+
+/datum/action/item_action/chameleon/change/id_trim/initialize_disguises()
+	// Little bit of copypasta but we only use trim datums rather than item paths
+	name = "Change [chameleon_name] Appearance"
+	build_all_button_icons()
+
+	LAZYINITLIST(chameleon_typecache)
+	LAZYINITLIST(chameleon_list)
+
+	for(var/datum/id_trim/trim_path as anything in typesof(/datum/id_trim))
+		if(chameleon_blacklist[trim_path])
+			continue
+
+		var/datum/id_trim/trim = SSid_access.trim_singletons_by_path[trim_path]
+
+		if(trim && trim.trim_state && trim.assignment)
+			var/chameleon_item_name = "[trim.assignment] ([trim.trim_state])"
+			chameleon_list[chameleon_item_name] = trim_path
+			chameleon_typecache[trim_path] = TRUE
+
+/datum/action/item_action/chameleon/change/id_trim/update_item(picked_trim_path)
+	var/obj/item/card/id/advanced/chameleon/agent_card = target
+
+	SSid_access.apply_trim_to_chameleon_card(agent_card, picked_trim_path, TRUE)
+
+	agent_card.update_label()
+	agent_card.update_appearance(UPDATE_ICON)
+
+/datum/action/item_action/chameleon/change/gun
+	chameleon_type = /obj/item/gun
+	chameleon_name = "Gun"
+
+/datum/action/item_action/chameleon/change/gun/New(Target)
+	chameleon_blacklist |= typecacheof(/obj/item/gun/energy/minigun)
+	. = ..()
+	if(!istype(target, /obj/item/gun/energy/laser/chameleon))
```

Comment:
i really detest having the parent call sandwiched in the middle of the proc in new code, could we implement a new proc called `setup_blacklist()` that we invoke on the `New()` in parent? Or is that just not worth it? I just don't like the fact that this could be a common pattern given how nice and clean the rest of this code is shaping up to be.